File: java/core/src/org/daisy/urakawa/xuk/SaveXukAction.java
package org.daisy.urakawa.xuk;

import java.io.IOException;
import java.net.URI;

import org.daisy.urakawa.command.CommandCannotExecuteException;
import org.daisy.urakawa.events.CancellableEvent;
import org.daisy.urakawa.events.Event;
import org.daisy.urakawa.events.EventHandler;
import org.daisy.urakawa.events.IEventHandler;
import org.daisy.urakawa.events.IEventListener;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.exception.MethodParameterIsOutOfBoundsException;
import org.daisy.urakawa.nativeapi.FileStream;
import org.daisy.urakawa.nativeapi.IStream;
import org.daisy.urakawa.nativeapi.IXmlDataWriter;
import org.daisy.urakawa.nativeapi.XmlDataWriter;
import org.daisy.urakawa.progress.ProgressAction;
import org.daisy.urakawa.progress.ProgressCancelledException;
import org.daisy.urakawa.progress.ProgressInformation;

/**
 *
 */
public class SaveXukAction extends ProgressAction implements
        IEventListener<CancellableEvent>
{
    protected IEventHandler<Event> mEventNotifier = new EventHandler();
    private URI mUri;
    private IXukAble mXukAble;
    private IStream mStream;
    private IXmlDataWriter mWriter;

    private static IStream getStreamFromUri(URI uri)
    {
        return new FileStream(uri.getPath());
    }

    private void initializeXmlWriter(IStream stream)
    {
        mWriter = new XmlDataWriter(stream);
    }

    /**
     * @param xukAble
     *        cannot be null
     * @param uri
     *        can be null
     * @param writer
     *        cannot be null
     * @throws MethodParameterIsNullException
     */
    public SaveXukAction(IXukAble xukAble, URI uri, IXmlDataWriter writer)
            throws MethodParameterIsNullException
    {
        if (xukAble == null || writer == null)
        {
            throw new MethodParameterIsNullException();
        }
        mUri = uri;
        mXukAble = xukAble;
        mWriter = writer;
        mStream = mWriter.getBaseStream();
    }

    /**
     * @param xukAble
     *        cannot be null
     * @param uri
     *        can be null
     * @param stream
     *        cannot be null
     * @throws MethodParameterIsNullException
     */
    public SaveXukAction(IXukAble xukAble, URI uri, IStream stream)
            throws MethodParameterIsNullException
    {
        if (xukAble == null || stream == null)
        {
            throw new MethodParameterIsNullException();
        }
        mUri = uri;
        mXukAble = xukAble;
        mStream = stream;
        initializeXmlWriter(mStream);
    }

    /**
     * @param xukAble
     *        cannot be null
     * @param uri
     *        cannot be null
     * @throws MethodParameterIsNullException
     */
    public SaveXukAction(IXukAble xukAble, URI uri)
            throws MethodParameterIsNullException
    {
        if (uri == null || xukAble == null)
        {
            throw new MethodParameterIsNullException();
        }
        mUri = uri;
        mXukAble = xukAble;
        mStream = getStreamFromUri(uri);
        initializeXmlWriter(mStream);
    }

    @Override
    public ProgressInformation getProgressInfo()
    {
        if (mStream == null)
        {
            return null;
        }
        ProgressInformation pi = null;
        try
        {
            pi = new ProgressInformation(mStream.getLength(), mStream
                    .getPosition());
        }
        catch (MethodParameterIsOutOfBoundsException e)
        {
            e.printStackTrace();
            return null;
        }
        return pi;
    }

    public boolean canExecute()
    {
        return mWriter != null;
    }

    private void closeOutput()
    {
        mWriter.close();
        mWriter = null;
        try
        {
            mStream.close();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
        mStream = null;
    }

    /**
     * @tagvalue Events "Cancelled-Finished"
     */
    public void execute() throws CommandCannotExecuteException
    {
        mCancelHasBeenRequested = false;
        mWriter.writeStartDocument();
        mWriter.writeStartElement("Xuk", IXukAble.XUK_NS);
        if (IXukAble.XUK_XSD_PATH != "")
        {
            if (IXukAble.XUK_NS.length() == 0)
            {
                mWriter.writeAttributeString("xsi",
                        "noNamespaceSchemaLocation",
                        "http://www.w3.org/2001/XMLSchema-instance",
                        IXukAble.XUK_XSD_PATH);
            }
            else
            {
                mWriter.writeAttributeString("xsi",
                        "noNamespaceSchemaLocation",
                        "http://www.w3.org/2001/XMLSchema-instance",
                        IXukAble.XUK_NS + " " + IXukAble.XUK_XSD_PATH);
            }
        }
        try
        {
            registerListener(this, CancellableEvent.class);
        }
        catch (MethodParameterIsNullException e1)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e1);
        }
        boolean canceled = false;
        try
        {
            mXukAble.xukOut(mWriter, mUri, this);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ?!", e);
        }
        catch (XukSerializationFailedException e)
        {
            closeOutput();
            throw new CommandCannotExecuteException(e);
        }
        catch (ProgressCancelledException e)
        {
            canceled = true;
        }
        finally
        {
            try
            {
                unregisterListener(this, CancellableEvent.class);
            }
            catch (MethodParameterIsNullException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ?!", e);
            }
            mWriter.writeEndElement();
            mWriter.writeEndDocument();
            closeOutput();
            if (canceled)
                notifyCancelled();
            else
                notifyFinished();
        }
    }

    public String getLongDescription()
    {
        return null;
    }

    public String getShortDescription()
    {
        return null;
    }

    @SuppressWarnings("unused")
    public void setLongDescription(String str)
            throws MethodParameterIsNullException
    {
        /**
         * Does nothing.
         */
    }

    @SuppressWarnings("unused")
    public void setShortDescription(String str)
            throws MethodParameterIsNullException,
            MethodParameterIsEmptyStringException
    {
        /**
         * Does nothing.
         */
    }

    public <K extends Event> void notifyListeners(K event)
            throws MethodParameterIsNullException
    {
        if (event == null)
        {
            throw new MethodParameterIsNullException();
        }
        mEventNotifier.notifyListeners(event);
    }

    public <K extends Event> void registerListener(IEventListener<K> listener,
            Class<K> klass) throws MethodParameterIsNullException
    {
        if (listener == null || klass == null)
        {
            throw new MethodParameterIsNullException();
        }
        mEventNotifier.registerListener(listener, klass);
    }

    public <K extends Event> void unregisterListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (listener == null || klass == null)
        {
            throw new MethodParameterIsNullException();
        }
        mEventNotifier.unregisterListener(listener, klass);
    }

    public <K extends CancellableEvent> void eventCallback(K event)
            throws MethodParameterIsNullException
    {
        if (event == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (event.isCancelled())
        {
            requestCancel();
        }
    }
}

File: csharp/core/xuk/SaveXukAction.cs
using System;
using System.IO;
using System.Xml;
using urakawa.command;
using urakawa.progress;

namespace urakawa.xuk
{

    ///<summary>
    ///  Action that serializes a xuk data stream from a <see cref="XukAble"/>
    ///</summary>
    public class SaveXukAction : ProgressAction
    {

        private Uri mDestUri;
        private Stream mDestStream;
        private XmlWriter mXmlWriter;
        private readonly IXukAble mSourceXukAble;

        private static Stream GetStreamFromUri(Uri src)
        {
            FileStream fs = new FileStream(src.LocalPath, FileMode.Create, FileAccess.Write);
            return fs;
        }

        private void initializeXmlWriter(Stream stream)
        {
            XmlWriterSettings settings = new XmlWriterSettings();
            settings.Indent = true;
            settings.IndentChars = " ";
            mXmlWriter = XmlWriter.Create(stream, settings);
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="destUri">The <see cref="Uri"/> of the destination (can be null)</param>
        /// <param name="xukAble">The source <see cref="IXukAble"/> (cannot be null)</param>
        /// <param name="writer">The destination <see cref="XmlWriter"/> (cannot be null)</param>
        public SaveXukAction(IXukAble xukAble, Uri destUri, XmlWriter writer)
        {
            if (writer == null)
                throw new exception.MethodParameterIsNullException(
                    "The destination Writer of the SaveXukAction cannot be null");
            if (xukAble == null)
                throw new exception.MethodParameterIsNullException(
                    "The source XukAble of the SaveXukAction cannot be null");
            mDestUri = destUri;
            mSourceXukAble = xukAble;
            mXmlWriter = writer;
            XmlTextWriter txtWriter = writer as XmlTextWriter;
            if (txtWriter != null)
            {
                mDestStream = txtWriter.BaseStream;
            }
        }
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="destUri">The <see cref="Uri"/> of the destination (can be null)</param>
        /// <param name="xukAble">The source <see cref="IXukAble"/> (cannot be null)</param>
        /// <param name="destStream">The destination <see cref="Stream"/> (cannot be null)</param>
        public SaveXukAction(IXukAble xukAble, Uri destUri, Stream destStream)
        {
            if (destStream == null)
                throw new exception.MethodParameterIsNullException(
                    "The destination Stream of the SaveXukAction cannot be null");
            if (xukAble == null)
                throw new exception.MethodParameterIsNullException(
                    "The source XukAble of the SaveXukAction cannot be null");
            mDestUri = destUri;
            mSourceXukAble = xukAble;
            mDestStream = destStream;
            initializeXmlWriter(mDestStream);
        }


        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="destUri">The <see cref="Uri"/> of the destination (cannot be null)</param>
        /// <param name="xukAble">The source <see cref="IXukAble"/>(cannot be null)</param>
        public SaveXukAction(IXukAble xukAble, Uri destUri)
        {
            if (destUri == null)
                throw new exception.MethodParameterIsNullException(
                    "The destination URI of the SaveXukAction cannot be null");
            if (xukAble == null)
                throw new exception.MethodParameterIsNullException(
                    "The source XukAble of the SaveXukAction cannot be null");

            mDestUri = destUri;
            mSourceXukAble = xukAble;
            mDestStream = GetStreamFromUri(mDestUri);
            initializeXmlWriter(mDestStream);
        }

        private void closeOutput()
        {
            mXmlWriter.Close();
            mXmlWriter = null;
            mDestStream.Close();
            mDestStream = null;
        }

        #region Overrides of ProgressAction

        /// <summary>
        /// Gets the current and estimated total progress values
        /// </summary>
        /// <param name="cur">A <see cref="long"/> in which the current progress is returned</param>
        /// <param name="tot">A <see cref="long"/> in which the estimated total progress is returned</param>
        protected override void getCurrentProgress(out long cur, out long tot)
        {
            if (mDestStream!=null)
            {
                cur = mDestStream.Position;
                tot = mDestStream.Length;
            }
            else
            {
                cur = 0;
                tot = 0;
            }
        }

        /// <summary>
        /// Gets a <c>bool</c> indicating if the <see cref="IAction"/> can execute
        /// </summary>
        /// <returns>The <c>bool</c></returns>
        public override bool canExecute()
        {
            return mXmlWriter != null;
        }


        /// <summary>
        /// Get a long uman-readable description of the command
        /// </summary>
        public override string getLongDescription()
        {
            return "Serializes a XUK fragment";
        }

        /// <summary>
        /// Execute the command.
        /// </summary>
        /// <exception cref="urakawa.exception.CannotExecuteException">Thrown when the command cannot be reversed.</exception>
        public override void execute()
        {
            mHasCancelBeenRequested = false;
            progress += SaveXukAction_progress;

            bool canceled = false;
            try
            {
// ReSharper disable PossibleNullReferenceException
                    mXmlWriter.WriteStartDocument();
// ReSharper restore PossibleNullReferenceException
                    mXmlWriter.WriteStartElement("Xuk", urakawa.ToolkitSettings.XUK_NS);
                    if (urakawa.ToolkitSettings.XUK_XSD_PATH != String.Empty)
                    {
                        if (urakawa.ToolkitSettings.XUK_NS == String.Empty)
                        {
                            mXmlWriter.WriteAttributeString(
                                "xsi", "noNamespaceSchemaLocation",
                                "http://www.w3.org/2001/XMLSchema-instance",
                                urakawa.ToolkitSettings.XUK_XSD_PATH);
                        }
                        else
                        {
                            mXmlWriter.WriteAttributeString(
                                "xsi",
                                "noNamespaceSchemaLocation",
                                "http://www.w3.org/2001/XMLSchema-instance",
                                String.Format("{0} {1}", urakawa.ToolkitSettings.XUK_NS, urakawa.ToolkitSettings.XUK_XSD_PATH));
                        }
                    }
                    mSourceXukAble.xukOut(mXmlWriter, mDestUri, this);
                    mXmlWriter.WriteEndElement();
                    mXmlWriter.WriteEndDocument();
            }
            catch (exception.ProgressCancelledException)
            {
                canceled = true;
            }
            finally
            {
                progress -= SaveXukAction_progress;

                closeOutput();

                if (canceled) notifyCancelled();
                else notifyFinished();
            }

        }

        private void SaveXukAction_progress(object sender, urakawa.events.progress.ProgressEventArgs e)
        {
            if (mHasCancelBeenRequested) e.Cancel();
        }

        /// <summary>
        /// Gets a short humanly readable description of the command
        /// </summary>
        public override string getShortDescription()
        {
            return "Save Xuk";
        }

        #endregion
    }
}
